Review episode 2: Resource Requirements (#20)
stderr-enst/jobefficiency: episodes/02_resourcerequirements.md

- The `normal` partition has a (relatively) small amount of memory and limits jobs to at most one day, but has by far the most nodes. This partition is probably designed for small- to medium-sized jobs.
  Since there are no `GRES` in this partition, only CPU computations can be performed here.

>   Also, as the number of cores per node is (relatively) small, this partition only allowed multithreading up to 36 threads on the same node. MPI is required for a higher degree of parallelism across multiple nodes.

Reviewer: I think my typo was "s -> d" instead of a forgotten "e" (since the rest is also in present tense).
```suggestion
  Also, as the number of cores per node is (relatively) small, this partition only allows multithreading up to 36 threads on the same node. MPI is required for a higher degree of parallelism across multiple nodes.
```

```bash
lscpu | grep "CPU(s):"
```
```output
CPU(s):                               192
NUMA node0 CPU(s):                    0-191
```

If `lscpu` is not available on your machine, you can also use
``` bash
cat /proc/cpuinfo | grep "core id" | wc -l
```
``` output
192
```

With $223\ \text{nodes} \times 36\ \text{cores}(+ \text{other partitions})$ CPU cores, the cluster has *multiple orders of magnitude* more cores in total than the login node or your local machine.

To see the amount of memory on the machine you are logged into you can use

```bash
cat /proc/meminfo | grep "MemTotal"
```
```output
MemTotal:       395695636 kB
```

Again, the total memory of the cluster is going to be much, much larger than the memory of any individual machine.

All of the CPU cores and memory are shared between you and all the other users of your cluster.
To get a feeling for the amount of resources per user, let's try to get an estimate for how many users there are by counting the number of home directories.

```bash
find /home -maxdepth 1 -mindepth 1 -type d | wc -l
```
```output

> 250

Reviewer: On our cluster it's more like 700, but I guess the concrete number does not matter much (also we need to decommission inactive users more often...).

:::

By dividing the total number of cores / the total memory by the amount of users, you get an estimate of how many resources each user has available if resources were equally distributed world.
For example, for 250 users we would approximately get:

|Partition   |CPU Cores / user|Memory / user|GRES / user|
|------------|----------------|-------------|-----------|
|`normal`    |32              |85 GiB       |-|
|`long`      |13              |69 GiB       |-|
|`express`   |1               |2 GiB        |-|
|`zen4`      |35              |141 GiB      |-|
|`gpuexpress`|0.128           |1 GiB        |0.028|
|`gpu4090`   |1               |12 GiB       |0.192|
|`gpuh200`   |2               |25 GiB       |0.128|

::: discussion

Do the per-user limits mean we can never use more than this amount of resources?
:::
::: solution
Users are typically not limited to the average per-user resources that we've calculated here.
In reality, we observe:

- The average is over a long periods of time. Short term you can usually use much more.

> - Users are in different phases of their projects, some may require computational resources now Others are still in preparation and need them later.

Reviewer: ```suggestion
- Users are in different phases of their projects, some may require computational resources while others are still in preparation and need them later.
```

mpirun -n 4 ./SnowmanRaytracer/build/raytracer -width=1024 -height=1024 -spp=256 -threads=1 -alloc_mode=3 -png=snowman.png
```

We define job resource requirements through the following `#SBATCH` directives (line-by-line):

- 1 node ...
- ... from the partition `<put your partition here>`
- 4 MPI tasks ...
- ... each of which uses one CPU core (so 4 cores in total)
- 1 GB of memory per node
- A timelimit of 1 minute

> The last two `#SBATCH` directives redirect the jobs output to the file `snowman-stdout.log` and assign the name "snowman" to the job.

Reviewer: If `-%j` is added at the top, it should be reflected here.
```suggestion
The last two `#SBATCH` directives redirect the jobs output to the file `snowman-stdout-<job id>.log` and assign the name "snowman" to the job.
```
I guess whether you add the job id to the log filename or not is a matter of taste. I find the log file spam annoying when tinkering with a job, so I like to overwrite the old log files. Other people might value more to keep old log files around for later comparison.

Finally, `Memory utilized` line shows the peak memory consumption that your job had at any point during its runtime, while `Memory Efficiency` is the ratio between this peak value and the requested amount of memory for the allocation. As we will see later, this value has to be taken with a grain of salt.

> Starting from the set of parameters that successfully run our program, we can now try to reduce the amount of requested resources. As is good scientific practice, we should only vary one parameter at a time and observe the result. Let's start by reducing the time limit. There is often a bit of variation in the time needed to run a job since not all nodes are perfectly identical, so you should add a arbitrarily scaled safety margin of maybe 10 percent. According to the time reported by `seff`, seven minutes should therefore be a good time limit. If your cluster is faster, you might reduce this even further. 

Reviewer: "Arbitrarily scaled" sounds like there is some base value you then scale randomly, which might be confusing. Maybe just "arbitrarily chosen" to keep things simple?
```suggestion
Starting from the set of parameters that successfully run our program, we can now try to reduce the amount of requested resources. As is good scientific practice, we should only vary one parameter at a time and observe the result. Let's start by reducing the time limit. There is often a bit of variation in the time needed to run a job since not all nodes are perfectly identical, so you should add an arbitrarily chosen safety margin of maybe 10 percent. According to the time reported by `seff`, seven minutes should therefore be a good time limit. If your cluster is faster, you might reduce this even further. 
```

This behavior seems contradictory at first: Slurm reported previously that our job only used around 367 MB of memory at most, which is well below the 500 MB limit we set. The explanation for this discrepancy lies in the fact that Slurm measures the peak memory consumption of jobs by *polling*, i.e., by periodically sampling how much memory the job currently uses. Unfortunately, if the program has spikes in memory consumption that are small enough to fit between two samples, Slurm will miss them and report an incorrect peak memory value. Spikes in memory usage are quite common, for example if your application uses short-lived subprocesses. Most annoyingly, many programs allocate a large chunk of memory right at the end of the computation to write out the results. In the case of the snowman raytracer, we encode the raw pixel data into a PNG at the end, which means we temporarily keep both the raw image and the PNG data in memory.

> ![](fig/slurm-memory-sampling.svg){alt="Slurm determines memory consuption by *polling*, i.e., periodically checking on the memory consumption of your job. If you job has a memory allocation profile with short spikes in memory usage, the value reported by `seff` can be incorrect. In particular, if the job gets cancelled due to memory exhaustion, you should not rely on the value reported by `seff` as it is likely significantly too low."}

Reviewer: I can't test this right now, but isn't the alt text what is shown when the image can't be loaded (or for people with impaired vision using a screen reader)? If so, it might make more sense to add a description of the image instead of just a copy of the paragraph below.

```suggestion
![](fig/slurm-memory-sampling.svg){alt="A graph plotting memory consumption of a job over time. The curve has an upwards trend with intermittent spikes. The periodic sampling of Slurm are shown as vertical lines intersecting the graph in regular intervals. While one of the memory spikes coincides with a sampling point, two others do not. One of them is near the start of the execution time and falls between two sample points. The other is at the very end of the execution time and occurs after the last sample point."}
```

```bash
echo -n "Total amount of memory used (in bytes): "
CGROUPPATH="$(cat /proc/self/cgroup | awk -F ':' '{print $3}')"
cat /sys/fs/cgroup/${CGROUPPATH}/memory.peak
```

Let's break down what each line does:

- The first line prints out a nice label for our peak memory output. We use `-n` to omit the usual newline that `echo` adds at the end of its output.

> - For the next part we need the so-called "cgroup path" of our job. To find out this path, we can use the `/proc/self/cgroup` file, which contains this path as the third entry of a colon-separated list. Therefore, we read the contents of this file (`cat`) and extract the third entry of the colon-separated list (`awk -F ':' '{print $3}'`). Since we do this in `$(...)`, Bash will place the output of these commands (i.e., the cgroup path) at this point.

Reviewer: ```suggestion
- For the next part we need the so-called "cgroup path" of our job. To find out this path, we can use the `/proc/self/cgroup` file, which contains this path as the third entry of a colon-separated list. Therefore, we read the contents of this file (`cat`) and extract the third entry of the colon-separated list (`awk -F ':' '{print $3}'`). Since we do this in `$(...)`, Bash will place the output of these commands (i.e., the cgroup path) in the variable `CGROUPPATH`.
```

> - The third line outputs the contents of a file (`cat`). The path of this file starts with `/sys/fs/cgroup`, which is a location where the Linux kernel exports all the cgroups v2 information as files.

Reviewer: ```suggestion
- The third line outputs the contents of a file (`cat`). The path of this file starts with `/sys/fs/cgroup`, which is a location where the Linux kernel exports all the cgroups v2 information as files, followed by the cgroup path of our job in the `CGROUPPATH` variable.
```

- The final part of the path is the information we actually want from the cgroup. In our case, we are interested in `memory.peak`, which contains the peak memory consumption of the cgroup. 

When you submit your job and look at the output once it finishes, you will find a line like this:

```output
[...]
Total amount of memory used (in bytes): 579346432
[...]
```

So even though Slurm reported our job to only use 367.28 MB of memory, we actually used nearly 600 MB! With this measurement we can make an informed decision on how to set the memory limit for our job:

```bash
#SBATCH --mem=700M
```

Run your job again with this limit to verify that it completes successfully.

::: callout
# Too tight memory limits can reduce performance!

Slurm enforces memory limits for all processes in a job.
Besides the applications memory demand (RSS - resident set size), this also includes caching mechanism of the Linux Kernel, e.g. for file I/O.

Too tight memory limits can cause too small caches next to the applications memory demand.
This in turn can severely reduce the jobs performance in some cases.

:::

> 

Reviewer: Can we demo that?
